settings_org: Fix visibility time of `Saved` state and `Discard` buttons.

pragatiagrawal31 opened the pull request:
This fixes the bug where the `Saved` state fades out almost instantly(that
is actually 300 ms, by default). Now, we fixed the bug and defined the
`fadeout_duration` for `Saved` state to be 500 ms.

Fixes: #11737.

Let me know if it still gets fade out quickly, then I'm thinking to change the time limit from 500 ms to 800 ms, or if to any other intermediate value.
Here are the two GIF's for the same
- **GIF1** : with fadeout_duration `500 ms`.
![mar-07-2019 22-51-02](https://user-images.githubusercontent.com/40331304/53976548-45174f00-412d-11e9-9786-d2125a1e20b5.gif)
- **GIF2** : with fadeout_duration `800 ms`
![mar-07-2019 23-18-51](https://user-images.githubusercontent.com/40331304/53977840-ea332700-412f-11e9-9fb5-8e30b92baa9b.gif)

timabbott commented:
@pragatiagrawal31 thanks for working on this!  But it doesn't work correctly.  I did some fiddling and noticed a few things:

* We have a setTimeout() before making changes inside the success handler `save_organization_settings`, which is the wrong place to wait 500ms -- change it to 50000 and you'll see it's waiting before changing "saving" to "saved"
* So we want to move that setTimeout to live in `change_save_button_state` instead.  However, doing that naively (as shown below) also does the wrong thing -- the "discard" button is still there while waiting for the timeout.
* So I think we want something like the below code, except that we need to hide the "discard" button as well in the `succeeded` case.

(I highly encourage testing with changing the timeouts to like 5 seconds so that it's really clear what's happening in each stage, since that reveals the bugs really clearly).

```
 exports.change_save_button_state = function ($element, state) {
-    var show_hide_element = function (state, fadeout_duration) {
+    var show_hide_element = function (state, fadeout_delay) {
         if (state === 'show') {
             $element.removeClass('hide').addClass('.show').fadeIn(300);
             return;
         }
-        if (fadeout_duration === undefined) {
-            fadeout_duration = 300;
-        }
-        $element.fadeOut(fadeout_duration);
+        setTimeout(function() {
+            $element.fadeOut(300);
+        }, fadeout_delay);
     };
     var $saveBtn = $element.find('.save-button');
     var $textEl = $saveBtn.find('.icon-button-text');
@@ -573,8 +572,8 @@ exports.change_save_button_state = function ($element, state) {
         $textEl.text(i18n.t("Save changes"));
         $saveBtn.attr("data-status", "failed");
     } else if (state === 'succeeded') {
-        show_hide_element('hide', 500);
         $textEl.text(i18n.t("Saved"));
+        show_hide_element('hide', 3000);
         $saveBtn.attr("data-status", "saved");
     }
 };
@@ -696,9 +695,7 @@ exports.build_page = function () {
             data: data,
             success: function (response_data) {
                 failed_alert_elem.hide();
-                setTimeout(function () {
-                    exports.change_save_button_state(save_btn_container, "succeeded");
-                }, 500);
+                exports.change_save_button_state(save_btn_container, "succeeded");
                 if (success_continuation !== undefined) {
                     success_continuation(response_data);
                 }
```

pragatiagrawal31 commented:
I did go through setTimeOut function, but I thought we might intentionally have it there to wait between "saving" and "saved" state, which actually doesn't make sense.
 - Now, I have made all the desired changes, where `Saved` state persists for 500 ms and also during that time `Discard` button is hidden.
- Similarly `Discard` button persists for 500 ms and visibility of `Save changes` is hidden during that.
- Added a commit to refactor `change_save_button_state`.

Here, is a **GIF**  showing the current updated behaviour : 

![Mar-09-2019 14-45-15](https://user-images.githubusercontent.com/40331304/54075780-32d21800-42c9-11e9-8535-6daec7376c4d.gif)

Let me know if we still want to change the fadeOut time.

timabbott commented:
@pragatiagrawal31 thanks for the updated version of this.  The "save changes" behavior now feels basically correct (though I think we want 800ms, not 500ms, for how long it stays before disappearing).  However, I don't understand the changes to what happens when you click "discard" -- your new behavior feels confusing to me.  

Having played with options a bit, I think what we want to have happen when you click "discard" is both "Save changes" and "discard" disappear together (probably immediately, without a fade at all, but fading together might be OK).  I also think it was more readable to have all the calls go through `exports.change_save_button_state` rather than having a separate "discard" function; we can just have the "discard" case return early if we don't want it to run the final `show_hide_element` code.  

Can you clean that up @pragatiagrawal31?  

@rishig FYI in case you have thoughts on whether we fade out the buttons over 300msor just disappear immediately when you click "discard".

pragatiagrawal31 commented:
I've updated the PR, changes include : 
- Fade out time has been changed from 500ms to `800 ms`.
- Added a small commit, to make node tests of `test_change_save_button_state` function in `settings_org` more readable.
- `Discard` button fades out along with ` Save changes` button(in `300ms` using `show_hide_element` function). 
I tried both methods, but personally felt this behaviour of `Discard` button to be more desirable, so currently added fading out in 300ms, whatever be the final consensus, let me know, I'll be updating the PR accordingly.
 
 **GIF** showing fading of `Discard` and `Save changes` button in `300 ms`.

![Mar-10-2019 14-59-22](https://user-images.githubusercontent.com/40331304/54083095-618dd400-4345-11e9-926b-33fd7503eb1c.gif)

 **GIF** showing the behaviour, where `Discard` and `Save changes` disappear immediately.

![Mar-10-2019 15-28-23](https://user-images.githubusercontent.com/40331304/54083426-31483480-4349-11e9-886f-93eb899b39b6.gif)

@timabbott FYI.

timabbott commented:
Merged as the commits ending with 9eed78433e096562d1a18f91813e04e2dd38d4aa, after tweaking the commit message of the second commit to add a bit more clarity.

Thanks @pragatiagrawal31!
